### PR TITLE
Add a new option for building with Apple bitcode support

### PR DESCRIPTION
**Pull request by nirbheek**
Normally, people would just pass `-fembed-bitcode` in `CFLAGS`, but this conflicts with `-Wl,-dead_strip_dylibs` and `-bundle`, so we need it as an option so that those can be quietly disabled.

### CODE REVIEW

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -302,6 +304,14 @@ def sanitizer_link_args(value):
     args = ['-fsanitize=' + value]
     return args
 
+def option_enabled(boptions, options, option):
+    try:
+        if option not in boptions:
+            return False
+        return options[option].value
+    except:
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__

[E722] do not use bare except'

**Comment by nirbheek** [comment]
To be clear, now you have to run meson with `-Db_bitcode` or set that in `default_options:` instead of setting the bitcode flags in CFLAGS and LDFLAGS.

**Comment by jpakkane** [comment]
This needs some documentation on what the flag is and where and when you'd want to use it. But where should be put that? I don't think we have a page for documenting compiler and base options. Maybe we should add one?

**Comment by nirbheek** [comment]
Yes, this needs documentation. Good point on the new page. I'll write a page for it, and hook it up to a test that ensures that new options are updated there too.

The PR is tagged as WIP because I was waiting for gstreamer's iOS CI to churn through all packages and archs with this patch.

**Comment by nirbheek** [comment]
Added docs and test!

**Comment by jpakkane** [comment]
Nice work. Maybe we should error out if asneeded and bitcode have conflicting values instead? At the very least it should print some sort of warning.

**Comment by nirbheek** [comment]
> Maybe we should error out if asneeded and bitcode have conflicting values instead? At the very least it should print some sort of warning.

I had that at first, and then I removed it because the bitcode option is incompatible with a very large number of linker flags. Search for `cannotBeUsedWithBitcode` here: https://opensource.apple.com/source/ld64/ld64-253.9/src/ld/Options.cpp.auto.html (I get 109 matches).

What we could do, is that if `-Db_bitcode` is enabled, we print one warning saying that it is incompatible with many options all of which will be silently disabled, and we could have a docs page that lists the options that we silently disable. At least this way things will be documented.

Erroring out is not necessarily useful because the incompatibility is with the *default options* of meson (which are nice-to-have, not compulsory), so we should continue to work without requiring an action by the user.

**Comment by jpakkane** [comment]
Makes sense.

File: docs/markdown/Builtin-options.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+| --option=value |
+| --option value |
+| -Doption=value |
+
+They can also be edited after setup using `meson configure`.
+
+Installation options are all relative to the prefix, except:
+
+* When the prefix is `/usr`: `sysconfdir` defaults to `/etc`, `localstatedir` defaults to `/var`, and `sharedstatedir` defaults to `/var/lib`
+8 When the prefix is `/usr/local`: `localstatedir` defaults to `/var/local`, and `sharedstatedir` defaults to `/var/local/lib`
```
Comment (jon-turney) [inline docs/markdown/Builtin-options.md]: I guess '8' -> '*' ?

File: docs/markdown/Builtin-options.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+| c_winlibs    | see below     | free-form comma-separated list           | Standard Windows libs to link against
+| cpp_args     |               | free-form comma-separated list           | C++ compile arguments to use
+| cpp_std      | none          | none, c++98, c++03, c++11, c++14, c++17, <br/>c++1z, gnu++03, gnu++11, gnu++14, gnu++17, gnu++1z | C++ language standard to use
+| cpp_debugstl | false         | true, false                              | C++ STL debug mode
+| cpp_eh       | sc            | none, a, s, sc                           | C++ exception handling type
+| cpp_winlibs  | see below     | free-form comma-separated list           | Standard Windows libs to link against
+
+The default values of `c_winlibs` and `cpp_winlibs` are in compiler-specific
+argument forms, but the libraries are: kernel32, user32, gdi32, winspool,
+shell32, ole32, oleaut32 uuid, comdlg32, advapi32
```
Comment (jon-turney) [inline docs/markdown/Builtin-options.md]: comma after oleaut32

**Comment by nirbheek** [comment]
@jpakkane does this look ok now?

**Comment by jpakkane** [comment]
LGTM!

**Pull request merged**